### PR TITLE
Dynamic AI Functions

### PR DESCRIPTION
**Pull request by ghoulslash**
This feature lets you assign custom AI functions via scripting to customize AI logic for specific battles. This could be useful for boss fights, totem bosses, etc that you want specialized AI logic for.

An example based on use case in voyager:
```
setdynamicaifunc AI_TagBattlePreferFoe
multi_2_vs_2 TRAINER_SIRIUS_NOVA_HYPERION_TAG, Text_NovaInsurgence_Arrival_Hyperion_Loss, TRAINER_SIRIUS_NOVA_DEIMOS_RECRUIT_TAG, Text_NovaInsurgence_Arrival_DeimosRecruit_Loss, TRAINER_SIRIUS_WHARF_TRITON_PARTNER, TRAINER_BACK_PIC_TRITON
```

where `AI_TagBattlePreferFoe` prioritizes attacking the opposite battler index

### CODE REVIEW

**Comment by AlexOn1ine** [comment]
Could you add the function `AI_TagBattlePreferFoe` as an example? Or any other function like that to have an example in the repo. Also any reason to make a callnative for this instead of using a normal flag?

**Comment by ghoulslash** [comment]
> Could you add the function `AI_TagBattlePreferFoe` as an example? Or any other function like that to have an example in the repo. Also any reason to make a callnative for this instead of using a normal flag? (Ig saving flags would be one)

Sure. The idea behind not using a flag is to be able to customize AI for specific trainers, e.g. maybe you want your gym leaders and E4 to have specialized AI, there aren't 12 ai flags available so this would help with that

**Comment by ghoulslash** [comment]
> Could you add the function `AI_TagBattlePreferFoe` as an example? Or any other function like that to have an example in the repo. Also any reason to make a callnative for this instead of using a normal flag? (Ig saving flags would be one)

done. ready for review

File: src/battle_ai_main.c [hunk 1]
```diff
@@ -86,7 +89,7 @@ static s32 (*const sBattleAiFuncTable[])(u32, u32, u32, s32) =
     [25] = NULL,                     // Unused
     [26] = NULL,                     // Unused
     [27] = NULL,                     // Unused
-    [28] = NULL,                     // Unused
+    [28] = AI_DynamicFunc,               // AI_FLAG_DYNAMIC_FUNC
```
Comment (AlexOn1ine) [inline src/battle_ai_main.c:92]:
```suggestion
    [AI_DynamicFunc,           // AI_FLAG_DYNAMIC_FUNC
```

File: src/battle_ai_main.c [hunk 2]
```diff
@@ -86,7 +89,7 @@ static s32 (*const sBattleAiFuncTable[])(u32, u32, u32, s32) =
     [25] = NULL,                     // Unused
     [26] = NULL,                     // Unused
     [27] = NULL,                     // Unused
-    [28] = NULL,                     // Unused
+    [AI_DynamicFunc,           // AI_FLAG_DYNAMIC_FUNC
```
Comment (AlexOn1ine) [inline src/battle_ai_main.c:92]:
```suggestion
    [28] = AI_DynamicFunc,          // AI_FLAG_DYNAMIC_FUNC
```

**Review by AlexOn1ine — APPROVED** [review]

**Comment by Pawkkie** [comment]
I've only just this and ghoul this is really really cool, I've been dedicated a few AI flags to "important" trainer battles where I want specific behaviour and didn't even catch this in the PR stage.

This is super sweet, thanks for sharing! :D

**Pull request merged**